File: leet_SLWD_1208_same_substr.cc
class Solution {
public:
    int equalSubstring(string s, string t, int maxCost)
    {
        vector<int(*)(string, string, int)> Solutions {
            Sliding_window,
        };
        return Solutions[0](s, t, maxCost);
    }

    static int Sliding_window(string s, string t, int maxCost)
    {
        int L = 0;
        int R = -1;
        int res = 0;
        int N = s.length();
        int windowcost = 0;
        while (++R < N)
        {
            windowcost += abs(s[R] - t[R]);
            while (windowcost > maxCost)
            {
                windowcost -= abs(s[L] - t[L]);
                ++L;
            }
            res = max(res, R - L + 1);
        }
        return res;
    }
};

File: leet_SLWD_1208_same_substr.py
class Solution:
    def equalSubstring(self, s: str, t: str, maxCost: int) -> int:
        return [
            Sliding_window,
        ][0](s, t, maxCost)

def Sliding_window(s: str, t: str, maxCost: int) -> int:
    N = len(t)
    L = 0
    windowcost = 0
    res = 0
    for R in range(N):
        diff = ord(s[R]) - ord(t[R])
        windowcost += abs(diff)
        while windowcost > maxCost:
            diff = ord(s[L]) - ord(t[L])
            windowcost -= abs(diff)
            L += 1
        res = max(res, R - L + 1)
    return res
